src/hipscat/catalog/pixel_tree.py

```diff
@@ -0,0 +1,117 @@
+        self._create_tree(partition_info_df)
+
+    def __len__(self):
+        pixel_count = 0
+        for _, order_pixels in self.pixels.items():
+            pixel_count += len(order_pixels)
+        return pixel_count
+
+    def contains(self, hp_order: int, hp_pixel: int) -> bool:
+        """Check if tree contains a node at a given order and pixel
+
+        Args:
+            hp_order: HEALPix order to check
+            hp_pixel: HEALPix pixel number to check
+
+        Returns:
+            True if the tree contains the pixel, False if not
+        """
+        return hp_order in self.pixels and hp_pixel in self.pixels[hp_order]
+
+    def get_node(self, hp_order: int, hp_pixel: int) -> PixelNode | None:
+        """Get the node at a given order and pixel
+
+        Args:
+            hp_order: HEALPix order to get
+            hp_pixel: HEALPix pixel number to get
+
+        Returns:
+            The PixelNode at the index, or None if a node does not exist
+        """
+        if self.contains(hp_order, hp_pixel):
+            return self.pixels[hp_order][hp_pixel]
+        return None
+
+    def _create_tree(self, partition_info_df: pd.DataFrame):
+        """Creates the tree by recursively creating parent nodes until reaching the root from
+        each leaf pixel
+
+        Validates the pixel structure as it does so by checking for duplicate pixels or pixels
+        defined at multiple orders
+        """
+        for _, row in partition_info_df.iterrows():
+            self._create_node_and_parent_if_not_exist(
+                row[self.METADATA_ORDER_COLUMN_NAME],
+                row[self.METADATA_PIXEL_COLUMN_NAME],
+                PixelNodeType.LEAF,
+            )
+
+    def _create_node_and_parent_if_not_exist(
+        self, hp_order: int, hp_pixel: int, node_type: PixelNodeType
+    ):
+        """Creates a node and adds to `self.pixels` in the tree, and recursively creates parent
+        node if parent does not exist"""
+        if self.contains(hp_order, hp_pixel):
+            raise ValueError(
+                "Incorrectly configured catalog: catalog contains duplicate pixels"
+            )
+
+        if hp_order == 0:
+            self._create_node(hp_order, hp_pixel, node_type, self.root_pixel)
+            return
+
+        parent_order = hp_order - 1
+        parent_pixel = hp_pixel >> 2
+        if not self.contains(parent_order, parent_pixel):
+            self._create_node_and_parent_if_not_exist(
+                parent_order, parent_pixel, PixelNodeType.INNER
+            )
+
+        parent = self.pixels[parent_order][parent_pixel]
+
+        if parent.node_type != PixelNodeType.INNER:
+            raise ValueError("Incorrectly configured catalog: catalog contains pixels defined at "
+                             "multiple orders")
+
+        self._create_node(hp_order, hp_pixel, node_type, parent)
+
+    def _create_node(self, hp_order, hp_pixel, node_type, parent):
```

Comment:
Any reason not to include the type annotations here?

Reply:
None except I forgot. Added now, thanks.